### DISCUSSION TITLE
Lettuce 6.2.2.RELEASE Auto-Instrumentation Error

### DISCUSSION THREAD

**Discussion by Taragos** [open-telemetry/opentelemetry-java-instrumentation#9343] [Q&A]
Hello,

I am currently trying to use the auto-instrumentation of the java agent to capture my lettuce Redis calls, but I am running into an error.

Whenever the lettuce client is trying to call the setex async command, the application runs into an error loop.

```
java.lang.ClassCastException: class java.lang.String cannot be cast to class [B (java.lang.String and [B are in module java.base of loader 'bootstrap')
```

The stack trace is sadly not very helpful, with the last recorded call pointing to an RedisCodec.encodeValue() function, which I can't debug since its breakpoint is never triggered.

Did anyone else come across this or a similar error?

**Comment by mateuszrzeszutek** [comment]
Hey @Taragos ,

This looks like a bug, can you [create an issue](https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/new/choose) for this? Please include a reproduction scenario, or, at the very least, the lettuce call that throws.

**Reply by Taragos** [reply]
Hello @mateuszrzeszutek,

I am not sure if I can provide enough information to justify an issue, since I can't share stack traces or similar information because of an NDA.

But by now, I did find the problematic code lines:

```java
if (argument instanceof KeyArgument) {
      KeyArgument keyArg = (KeyArgument) argument;
      return stringCodec.decodeValue(keyArg.codec.encodeValue(keyArg.key));
    }
```

This is from https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/lettuce/core/protocol/OtelCommandArgsUtil.java 

What strikes me as odd, is that we are using the codec function `encodeValue `to encode the key, even if the interface has a method called `encodeKey`.

In my case I have a RedisCodec<String, byte[]> so that when the instrumentation is trying to use the `encodeValue `method to encode the String-key it obviously runs into an error, since the method expects an array of bytes.

The `encodeKey`-method would seemingly work in this case.

**Reply by mateuszrzeszutek** [reply]
That's a good find, thanks!
This method was supposed to work exactly like `KeyArgument`/`ValueArgument` `toString()` methods, unfortunately we've made a mistake with that. I fixed that in #9347 -- if you are able to do that, you could test the javaagent's snapshot release ~tomorrow and check if that helped.

**Reply by trajano** [reply]
Does that mean the automatic instrumentation should work with Lettuce Core 6.x?   Or does it only go to 5.x because I only see up to 5.x here

https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/lettuce

**Reply by mateuszrzeszutek** [reply]
It should work with 6.x.
In general, the instrumentation module names only contain the minimum supported version; the table [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md#libraries--frameworks) lists the full information about supported versions and in some cases you can find limitations there.

**Reply by trajano** [reply]
Unfortunately I don't see it working on a [simple app](https://github.com/trajano/simple-auth) that uses Spring Session with Redis data store

**Reply by mateuszrzeszutek** [reply]
Please open a new bug, describe what you expected and what actually happens, and add a link to your app.

**Reply by trajano** [reply]
https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/9812
